sqlmesh/lsp/main.py

```diff
@@ -31,14 +32,85 @@ def __init__(
         """
         self.server = LanguageServer(server_name, version)
         self.context_class = context_class
-        self.context: t.Optional[Context] = None
+        # A tuple of (context, reverse_map) where the reverse_map is uri to model name
+        self.context_and_reverse_map: t.Optional[t.Tuple[Context, t.Dict[str, str]]] = None
+        self.lint_cache: t.Dict[str, t.List[RuleViolationWithModelAndType]] = {}
 
         # Register LSP features (e.g., formatting, hover, etc.)
         self._register_features()
 
     def _register_features(self) -> None:
         """Register LSP features on the internal LanguageServer instance."""
 
+        @self.server.feature(types.TEXT_DOCUMENT_DID_OPEN)
+        def did_open(ls: LanguageServer, params: types.DidOpenTextDocumentParams) -> None:
+            if self.context_and_reverse_map is None:
+                self.ensure_context_for_document(
+                    ls.workspace.get_document(params.text_document.uri)
+                )
+            if self.context_and_reverse_map is None:
+                raise RuntimeError("No context found")
+            model = self.context_and_reverse_map[1][params.text_document.uri]
+            if model is None:
+                return
+            self.lint_cache[params.text_document.uri] = self.context_and_reverse_map[0].lint_models(
+                [model],
+                raise_on_error=False,
+            )
+            ls.publish_diagnostics(
+                params.text_document.uri,
+                SQLMeshLanguageServer._diagnostics_to_lsp_diagnostics(
+                    self.lint_cache[params.text_document.uri]
+                ),
+            )
+
+        @self.server.feature(types.TEXT_DOCUMENT_DID_CHANGE)
+        def did_change(ls: LanguageServer, params: types.DidChangeTextDocumentParams) -> None:
+            if self.context_and_reverse_map is None:
+                self.ensure_context_for_document(
+                    ls.workspace.get_document(params.text_document.uri)
+                )
+            if self.context_and_reverse_map is None:
+                raise RuntimeError("No context found")
```

Comment:
This looks like an exact copy paste of 46-51. I feel like this should be handled with something like:
```
def context_get_or_load()
```